src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -1177,46 +1185,70 @@ std::shared_ptr<ExpressionResult<Result>> ExpressionEvaluator::_evaluate_binary_
           Functor{}(values[row_idx], left.values.front(), right.values[row_idx]);
         }
       }
-
-      result = std::make_shared<ExpressionResult<Result>>(std::move(values), std::move(nulls));
     } else {
       Fail("BinaryOperation not supported on the requested DataTypes");
     }
   });
 
-  return result;
+  return std::make_shared<ExpressionResult<Result>>(std::move(values), std::move(nulls));
 }
 
 template <typename Result, typename Functor>
 std::shared_ptr<ExpressionResult<Result>> ExpressionEvaluator::_evaluate_binary_with_functor_based_null_logic(
     const AbstractExpression& left_expression, const AbstractExpression& right_expression) {
-  auto result = std::shared_ptr<ExpressionResult<Result>>{};
+  std::vector<Result> values;
+  std::vector<bool> nulls;
 
-  _resolve_to_expression_result_views(left_expression, right_expression, [&](const auto& left, const auto& right) {
+  _resolve_to_expression_results(left_expression, right_expression, [&](const auto& left, const auto& right) {
     using LeftDataType = typename std::decay_t<decltype(left)>::Type;
     using RightDataType = typename std::decay_t<decltype(right)>::Type;
 
     if constexpr (Functor::template supports<Result, LeftDataType, RightDataType>::value) {
-      const auto result_row_count = _result_size(left.size(), right.size());
-
-      std::vector<bool> nulls(result_row_count);
-      std::vector<Result> values(result_row_count);
+      const auto result_size = _result_size(left.size(), right.size());
+      values.resize(result_size);
+      nulls.resize(result_size);
+
+      auto get_null = [](const auto& result, ChunkOffset row_idx) -> bool {
```

Comment:
How big is the perf-impact, compared to the compile time saving? I think I instantiated all Literal/Segment/Nullable combinations here, intentionally.

Reply:
I'll measure it once nemea is idle, but my understanding is that we still initiate every combination here. It's just that we don't work our ways through layers and layers of templates to reach that.

Reply:
> but my understanding is that we still initiate every combination here

you're now checking for the nullability of an operand for every row, instead of instantiating this for every combination of nullable/non-nullable. TPC-H will not show you whether this had an impact, as this particular path is only taken for OR and ANDs and TPC-H is pretty light on those.

Reply:
> It's just that we don't work our ways through layers and layers of templates to reach that.

There is one layer that you removed here: That were the ExpressionResults are resolved to Views.

My issue is that this is more verbose now, and potentially slower, though likely faster to compile. But this is a path only taken for AND and OR, so I am not sure how big the compile time win really is to justify the verbosity and deviation from the intended way operations should be evaluated in the ExpressionEvaluator (which is: Resolve operands to views and implement you operation not caring about whether an operand is a Literal, a Series, Nullable or Not Nullable.)

Reply:
The performance difference is given in the PR description.

Not sure if I can follow though. How is this different from `_evaluate_binary_with_default_null_logic`?

Reply:
Aaah, ok, got it.

Reply:
Ok, can't come up with a good solution for now. Let's just look at the other things.